src/main/java/com/griddynamics/cloud/learning/service/UserService.java

```diff
@@ -0,0 +1,80 @@
+import com.griddynamics.cloud.learning.dao.repository.RoleRepository;
+import com.griddynamics.cloud.learning.dao.repository.UserRepository;
+import com.griddynamics.cloud.learning.web.dto.UserDto;
+import org.springframework.security.core.GrantedAuthority;
+import org.springframework.security.core.authority.SimpleGrantedAuthority;
+import org.springframework.security.core.userdetails.UserDetails;
+import org.springframework.security.core.userdetails.UserDetailsService;
+import org.springframework.security.core.userdetails.UsernameNotFoundException;
+import org.springframework.security.crypto.password.PasswordEncoder;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Optional;
+import java.util.Set;
+
+@Service
+public class UserService implements UserDetailsService {
+
+    private UserRepository userRepository;
+
+    private RoleRepository roleRepository;
+
+    private PasswordEncoder encoder;
+
+    public UserService(UserRepository userRepository, RoleRepository roleRepository, PasswordEncoder encoder) {
+        this.userRepository = userRepository;
+        this.roleRepository = roleRepository;
+        this.encoder = encoder;
+    }
+
+    @Override
+    @Transactional
+    public UserDetails loadUserByUsername(String username) throws UsernameNotFoundException {
+        User user = this.userRepository.findUserByUsername(username).orElseThrow(
+                () -> new UsernameNotFoundException("No user found with username " + username));
+
+        Set<GrantedAuthority> authorities = new HashSet<>();
+        for (Permission permission : user.getRole().getPermissions()) {
+            authorities.add(new SimpleGrantedAuthority(permission.toString()));
+        }
+
+        return new org.springframework.security.core.userdetails.User(user.getUsername(), user.getPassword(), authorities);
+    }
+
+    @Transactional
+    public User save(UserDto user) {
+
+        Optional<Role> existingRole = roleRepository.findAllBy().stream()
+                .filter(r -> r.getPermissions().size() == 1
+                        && r.getPermissions().contains(Permission.PURCHASE_ARTICLE)).findFirst();
+
+        Role userRole = existingRole.orElseGet(this::createUserRole);
+
+        User newUser = User.builder()
+                .email(user.getEmail())
+                .username(user.getUsername())
+                .password(encoder.encode(user.getPassword()))
+                .role(userRole)
+                .build();
+
+        return this.userRepository.save(newUser);
+    }
+
+    private Role createUserRole() {
+        Role newUserRole = Role.builder()
+                .name("AUTOGENERATED USER")
+                .permissions(Set.of(Permission.PURCHASE_ARTICLE))
+                .users(Collections.emptySet())
+                .build();
+
+        return roleRepository.save(newUserRole);
```

Comment:
Please don't persist it here

Reply:
Should it be implemented via a private method or something completely different?

Reply:
Check our previous service